feat(compiler): Add a TSConfig option `useTemplatePipeline`

dylhunn opened the pull request:
The Template Pipeline is a brand new backend for the Angular compiler, replacing `TemplateDefinitionBuilder` and friends. It generates the Ivy instructions corresponding to an input template (or host binding). The Template Pipeline has an all-new design based on an intermediate representation compiled over many phases, which will allow us to experiment with compiler changes more easily in the future.

With this commit, the template pipeline can now be enabled in any project via the `useTemplatePipeline` TSConfig option. However, it is still disabled by default.

alxhub commented on packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:
```diff
@@ -431,4 +431,9 @@ export interface MiscOptions {
   forbidOrphanComponents?: boolean;
+
+  /**
+   * Whether to use TemplateDefinitionBuilder as the code generator, or Template Pipeline.hand
```
`hand`? ✋

dylhunn replied:
I thought this comment might come in _handy_

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -356,7 +357,10 @@ function convertDirectiveFacadeToMetadata(facade: R3DirectiveMetadataFacade): R3
-    host: extractHostBindings(facade.propMetadata, facade.typeSourceSpan, facade.host),
+    host: {
+      ...extractHostBindings(facade.propMetadata, facade.typeSourceSpan, facade.host),
+      useTemplatePipeline: false,
```
We should probably introduce a constant for these (`SHOULD_USE_TEMPLATE_PIPELINE_FOR_JIT`?) so we don't forget to update them.

dylhunn committed:
feat(compiler): Add a TSConfig option `useTemplatePipeline`

The Template Pipeline is a brand new backend for the Angular compiler, replacing `TemplateDefinitionBuilder`. It generates the Ivy instructions corresponding to an input template (or host binding). The Template Pipeline has an all-new design based on an intermediate representation compiled over many phases, which will allow us to experiment with compiler changes more easily in the future.

With this commit, the template pipeline can now be enabled in any project via the `useTemplatePipeline` TSConfig option. However, it is still disabled by default.

dylhunn commented:
caretaker: the current two approvals are sufficient to merge.

jessicajaniuk commented:
This PR was merged into the repository by commit 47e6e841016abfca0c1aa84051d82a04b3027617.
